presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java

```diff
@@ -61,23 +74,84 @@ public final class SessionPropertyManager
+    {
+        this.systemSessionPropertyProviders.put("java-worker", new JavaWorkerSystemSessionPropertyProvider(javaFeaturesConfig, nodeSpillConfig));
         addSystemSessionProperties(systemSessionProperties);
     }
 
+    @Inject
+    public SessionPropertyManager(
+            SystemSessionProperties systemSessionProperties,
+            JavaFeaturesConfig javaFeaturesConfig,
+            FeaturesConfig featuresConfig,
+            NodeSpillConfig nodeSpillConfig,
+            ServerConfig serverConfig,
+            FunctionAndTypeManager functionAndTypeManager,
+            NodeManager nodeManager)
+    {
+        this(systemSessionProperties.getSessionProperties(), javaFeaturesConfig, nodeSpillConfig);
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+        this.nativeExecution = requireNonNull(featuresConfig, "featuresConfig is null").isNativeExecutionEnabled();
+        this.coordinatorSidecar = requireNonNull(serverConfig, "serverConfig is null").isCoordinatorSidecarEnabled();
+    }
+
+    // todo:  this hack is required to load "native" session property provider
+    public void loadSessionPropertyProvider()
+    {
+        if (nativeExecution && coordinatorSidecar) {
+            loadSessionPropertyProvider("native", ImmutableMap.of(), functionAndTypeManager, nodeManager);
+        }
+    }
+
+    public void loadSessionPropertyProvider(
+            String sessionPropertyProviderName,
+            Map<String, String> properties,
+            TypeManager typeManager,
+            NodeManager nodeManager)
+    {
+        requireNonNull(sessionPropertyProviderName, "sessionPropertyProviderName is null");
+        SystemSessionPropertyProviderFactory factory = systemSessionPropertyProviderFactories.get(sessionPropertyProviderName);
+        checkState(factory != null, "No factory for session property provider factory %s", sessionPropertyProviderName);
+        SystemSessionPropertyProvider sessionPropertyProvider = factory.create(properties,
+                new SessionPropertyContext(typeManager, nodeManager));
+
+        if (systemSessionPropertyProviders.putIfAbsent(sessionPropertyProviderName, sessionPropertyProvider) != null) {
+            throw new IllegalArgumentException(format("System session property provider is already registered for property provider [%s] ", sessionPropertyProviderName));
+        }
+    }
+
+    public void addSessionPropertyProviderFactory(SystemSessionPropertyProviderFactory factory)
+    {
+        if (systemSessionPropertyProviderFactories.putIfAbsent(factory.getName(), factory) != null) {
+            throw new IllegalArgumentException(format("System Session property provider factory '%s' is already registered", factory.getName()));
```

Comment:
string concatenation is simpler and less bug prone, especially when the placeholder is the final thing in the string